telegram/ext/filters.py

```diff
@@ -850,29 +851,68 @@ class user(BaseFilter):
         Examples:
             ``MessageHandler(Filters.user(1234), callback_method)``
 
+        Attributes:
+            user_ids(set(:obj:`int`), optional): Which user ID(s) to allow through.
+            usernames(set(:obj:`str`), optional): Which username(s) (without leading '@') to allow
+                through.
         Args:
-            user_id(:obj:`int` | List[:obj:`int`], optional): Which user ID(s) to allow through.
-            username(:obj:`str` | List[:obj:`str`], optional): Which username(s) to allow through.
-                If username starts with '@' symbol, it will be ignored.
+            user_id(:obj:`int` | List[:obj:`int`], optional): Which user ID(s) to allow
+                through.
+            username(:obj:`str` | List[:obj:`str`], optional): Which username(s) to allow
+                through. If username starts with '@' symbol, it will be ignored.
 
         Raises:
             ValueError: If chat_id and username are both present, or neither is.
 
         """
 
         def __init__(self, user_id=None, username=None):
-            if not (bool(user_id) ^ bool(username)):
+            if (user_id is None) == (username is None):
                 raise ValueError('One and only one of user_id or username must be used')
-            if user_id is not None and isinstance(user_id, int):
-                self.user_ids = [user_id]
-            else:
-                self.user_ids = user_id
-            if username is None:
-                self.usernames = username
-            elif isinstance(username, string_types):
-                self.usernames = [username.replace('@', '')]
-            else:
-                self.usernames = [user.replace('@', '') for user in username]
+
+            self._user_ids_lock = Lock()
+            self._usernames_lock = Lock()
+
+            # Initialize in a way that will not fail the first setter calls
+            self._user_ids = user_id
+            self._usernames = username
```

Comment:
Maybe initialize both to sets, saving some extra validation further down the line?

Correct me if I'm wrong, but for the usecase of this filter, a None value and an empty set has equivalent meaning.

Having this filter with two empty sets (usernames, user_ids) will mean it will allow all messages (my suggestion. you may decide exactly the opposite. you can even suggest controlling it with a variable)

```suggestion
            if isinstance(user_id, int):
                user_id = {user_id}
            elif not isinstance(user_id, set):
                user_id = set(user_id)
            self._user_ids = user_id
            
            # and similarly for `self._usernames`, `usernames` and compare to str & set
```

p.s. you should unify the if...elif....actually_set code with the property setters down below

Reply:
I tried to incorporate most of your suggestions. As to the ``None`` question:

The way this filter was designed before you could only user *either* usernames *or* user_ids. I guess because it's not clear what do to, if a user is allowed by username but not by id or vice versa. Personally, I wouldn't wanna change that. But that's why from my perspective ``None`` is not equal to an empty set here. So I kept if ``None`` checks.

Reply:
1. I understand why not have both types (usernames/groups).
2. The only question is what considered an as no value: None, empty set or both?

I claim that an empty set means: no value was set at all.

Where is a potential problem? When both usernames and userids are empty sets.
However, this is now an allowed scenario: set may start empty, or will be cleared further down the line.

So how do I suggest to approach this:
In the filter itself perform following checks:

```
if self._usernames:
   return filter_by_username()
if self._user_ids:
   Return filter_by_user_ids()
Return default_value

```
Default value can be denied for backward compat but we should allow control to the user to use a different default value (allow)

Reply:
Okay, I think I get it now. I'll update.